File: algorithms/java/45_jump_game_2/JumpGame.java
class JumpGame {
	// dp
	public int jump1(int[] nums) {
        if (nums == null || nums.length == 0) return Integer.MAX_VALUE;
        int[] m = new int[nums.length];
        Arrays.fill(m, Integer.MAX_VALUE);
        m[0] = 0;
        for (int i = 1; i < m.length; i++) {
            for (int j = 0; j < i; j++) {
                if (m[j] != Integer.MAX_VALUE && j + nums[j] >= i) {
                    m[i] = Math.min(m[i], m[j] + 1);
                }
            }
        }
        return m[m.length - 1];
    }

	// bfs
	public int jump2(int[] nums) {
		if (nums == null || nums.length == 0) return Integer.MAX_VALUE;
		if (nums.length == 1) return 0;
		Queue<Integer> queue = new ArrayDeque<>();
		boolean[] visited = new boolean[nums.length];
		queue.offer(0);
		visited[0] = true;
		int level = 0;
		while (!queue.isEmpty()) {
			int size = queue.size();
			for (int i = 0; i < size; i++) {
				Integer curIndex = queue.poll();
				for (int j = Math.min(curIndex + nums[curIndex], nums.length - 1); j >= curIndex; j--) {
					if (j == nums.length - 1) return level + 1;
					if (!visited[j]) {
						queue.offer(j);
						visited[j] = true;
					}
				}
			}
			level++;
		}
		return Integer.MAX_VALUE;
	}
}

File: algorithms/cpp/45_jump_game2/JumpGame.cpp
class JumpGame {
public:
	// 1. dp
	int jump_dp(vector<int>& nums) {
		vector<int> dp(nums.size());
		dp[dp.size() - 1] = 0;
		for (int i = num.size() - 2; i >= 0; --i) {
			int maxStep = min(i + nums[i], (int)nums.size() - 1);
			dp[i] = INT_MAX;
			for (int j = i + 1; j <= maxStep; ++j) {
				if (dp[i] != INT_MAX) {
					dp[i] = min(dp[i], dp[j] + 1);
				}
			}
		}
		return dp[0];
	}

	// 2. bfs
	int jump_bfs(vector<int>& nums) {
		if (nums.size() == 0)
            return INT_MAX;
        if (nums.size() == 1)
            return 0;
        queue<int> queue;
        vector<bool> visited(nums.size());
        queue.push(0);
        visited[0] = true;
        int steps = 0;
        while (!queue.empty()) {
            int size = queue.size();
            for (int i = 0; i < size; ++i) {
                int cur_index = queue.front();
                queue.pop();
                for (int j = min(cur_index + nums[cur_index], (int)nums.size() - 1); j >= cur_index; --j) {
                    if (j == nums.size() - 1)
                        return steps + 1;
                    if (!visited[j]) {
                        queue.push(j);
                        visited[j] = true;
                    }
                }
            }
            ++steps;
        }
        return INT_MAX;
	}
};
